### PR TITLE
feature(PrototypeRouteListener): add support to set default_version per route

### PR DESCRIPTION
**Pull request by solcre**
Hey @weierophinney ,

This pull request allows to set the default version per route. It is useful for Apigility where you can't set the version per API. 

Examples and Test Cases submitted. 

Regards!

Matías Fuster

### CODE REVIEW

**Comment by solcre** [comment]
This provides a solution to: [Apigility Issue 164](https://github.com/zfcampus/zf-apigility/issues/164)

File: README.md [hunk 1]
```diff
@@ -97,14 +97,26 @@ would look like:
...
 'zf-versioning' => [
-    'default_version' => 1,
+    'default_version' => 2,
+],
+```
+Or
+```php
+// Set default version to v2 and v3 for the users and status routes respectively
+'zf-versioning' => [
+    'default_version' => array(
```
Comment (webimpress) [inline README.md]: Please use short array notation here `[`. Don't forget to update also closing bracket.

File: README.md [hunk 2]
```diff
@@ -97,14 +97,26 @@ would look like:
...
+        'myapi.rest.users' => 2,
+        'myapi.rpc.status' => 3,
+    ),
 ],
 ```
-
```
Comment (webimpress) [inline README.md]: Let's revert this change, would be nice to have an empty line before next section.

File: src/PrototypeRouteListener.php [hunk 3]
```diff
@@ -117,9 +119,16 @@ public function onMergeConfig(ModuleEvent $e)
                     . $config['router']['routes'][$routeName]['options']['route'];
             }
 
+            $routeVersion = $this->versionRouteOptions;
+            if (isset($config['zf-versioning']['default_version']) &&
+            is_array($config['zf-versioning']['default_version']) &&
+            isset($config['zf-versioning']['default_version'][$routeName])) {
```
Comment (webimpress) [inline src/PrototypeRouteListener.php]: We can simplify the condition here, the last `isset` is going to be enough:
```php
if (isset($config['zf-versioning']['default_version'][$routeName])) {
```

File: test/PrototypeRouteListenerTest.php [hunk 4]
```diff
@@ -136,4 +138,98 @@ public function testPrototypeAddedToRoutesProvidedToListener(array $routes, $api
             $this->assertEquals($apiVersion, $options['defaults']['version']);
         }
     }
+    public function defaultVersionValues()
```
Comment (webimpress) [inline test/PrototypeRouteListenerTest.php]: Please add empty line before new method.

**Comment by solcre** [comment]
@webimpress Changes made!

File: src/PrototypeRouteListener.php [hunk 5]
```diff
@@ -117,9 +119,15 @@ public function onMergeConfig(ModuleEvent $e)
                     . $config['router']['routes'][$routeName]['options']['route'];
             }
 
+            $routeVersion = $this->versionRouteOptions;
+            if (isset($config['zf-versioning']['default_version']) &&
```
Comment (webimpress) [inline src/PrototypeRouteListener.php]: As I noted before the last `isset` check is sufficient and the top one can be removed.

Comment (solcre) [reply src/PrototypeRouteListener.php]: ouch! done now!

File: src/PrototypeRouteListener.php [hunk 6]
```diff
@@ -83,7 +83,9 @@ public function onMergeConfig(ModuleEvent $e)
         }
 
         // Override default version of 1 with user-specified config value, if available.
-        if (isset($config['zf-versioning']['default_version'])) {
+        if (isset($config['zf-versioning']['default_version'])
+            && !is_array($config['zf-versioning']['default_version'])
```
Comment (weierophinney) [inline src/PrototypeRouteListener.php]: This should probably be `is_scalar` instead of `! is_array`.

(I recommend `is_scalar` and not `is_numeric`, as we may allow string values at some point: e.g., to accommodate semantic versioning, or to allow "project name" versioning, ala the GitHub API.)

File: README.md [hunk 7]
```diff
@@ -97,11 +97,24 @@ would look like:
 The `default_version` key provides the default version number to use in case a version is not
 provided by the client.  `1` is the default for `default_version`.
 
+It accepts an `integer` to set the default version number to *all* routes or an associative `array` with the route name as key and the version number as value to set a specific version for each route.
+
```
Comment (weierophinney) [inline README.md]: This should be rephrased to better differentiate the two setting types. May I suggest the following?

```markdown
The setting accepts one of the two following possible values:

- A PHP `integer` indicating the default version number for *all* routes.
- An associative array, where the keys are route names, and the values the default version to use with the associated route.
```

File: README.md [hunk 8]
```diff
@@ -97,11 +97,24 @@ would look like:
...
 Full Example:
 
 ```php
+// Set v2 as default version for all routes
+'zf-versioning' => [
+    'default_version' => 2,
+],
+```
+Or
+```php
```
Comment (weierophinney) [inline README.md]: Add a blank line before and after the line containing `Or`, please.

File: src/PrototypeRouteListener.php (hunk 6 above)
Comment (solcre) [reply src/PrototypeRouteListener.php]: much better for sure! Done with the changes!

File: README.md (hunk 7 above)
Comment (solcre) [reply README.md]: Thanks for helping me improve my communication in english, was a bit messy to be honest. heh

**Comment by solcre** [comment]
@weierophinney all done!

**Comment by weierophinney** [comment]
Thanks, @solcre!
